docs/source/internals/engine.rst

```diff
@@ -1,11 +1,49 @@
-.. todo::
+.. _internal_architecture:engine:
 
-    .. _internal_architecture:engine:
+******
+Engine
+******
 
-    ******
-    Engine
-    ******
 
-    `#4038`_
+
+.. _internal_architecture:engine:caching:
+
+Controlling caching
+-------------------
+
+.. important::
+
+    This section covers some details of the caching mechanism which are not discussed in the :ref:`topics section <topics:provenance:caching>`.
+    If you are developing plugins and want to modify the caching behavior of your classes, we recommend you read that section first.
+
+There are several methods which the internal classes of AiiDA use to control the caching mechanism:
+
+On the level of the generic :class:`orm.Node <aiida.orm.nodes.Node>` class:
+
+* The :meth:`~aiida.orm.nodes.Node.is_valid_cache` property determines whether a particular node can be used as a cache.
+  This is used for example to disable caching from failed calculations.
+* Node classes have a ``_cachable`` attribute, which can be set to ``False`` to completely switch off caching for nodes of that class.
+  This avoids performing queries for the hash altogether.
+
+On the level of the :class:`Process <aiida.engine.processes.process.Process>` and :class:`orm.ProcessNode <aiida.orm.nodes.process.ProcessNode>` classes:
+
+* The :meth:`ProcessNode.is_valid_cache <aiida.orm.nodes.process.ProcessNode.is_valid_cache>` calls :meth:`Process.is_valid_cache <aiida.engine.processes.process.Process.is_valid_cache>`, passing the node itself.
```

Comment:
Since this is a way for plugin developers to override the caching behavior, shouldn't this be mentioned in the section for plugin developers?

Reply:
I interpret that you mean it in the sense that "...shouldn't this ALSO be mentioned...", right? Since here it is explaining the internal mechanisms by which the feature is implemented (in the same way the 3rd point below this explains where the `exit_codes` with `invalidate_cache` are processed). If that is the case, yes, I agree there should be also a more "application oriented" reference to this in the topics.

Reply:
Yep, _also_ mentioned in the topics.